sunpy/data/data_manager/cache.py

```diff
@@ -73,23 +73,33 @@ def download(self, urls, namespace='', redownload=False):
             if details:
                 break
         if details:
-            if redownload or self._has_expired(details):
-                # if file is in cache and it has to be redownloaded or the cache has expired
+            if not (redownload or self._has_expired(details)):
+                return Path(details['file_path'])
+        try:
+            if(details and redownload):
+                # if file is in cache and it has to be redownloaded
                 # then remove the file and delete the details from the storage
                 os.remove(details['file_path'])
                 self._storage.delete_by_key('url', details['url'])
-            else:
-                return Path(details['file_path'])
 
-        file_path, file_hash, url = self._download_and_hash(urls, namespace)
+            file_path, file_hash, url = self._download_and_hash(urls, namespace)
```

Comment:
I think to achieve the aims of the issue this line needs to be first under the try block. You need to do the download first, and then if the download fails you need a new code path which returns the cached version. I think from looking at your code at the moment you can still end up in a situation where you delete the file and the download fails.

Reply:
I've made a few changes you can review them.
I think I've made a lot of uneccessory commits for a single PR, you can ask if you want me to make a new PR and close this.

Reply:
We can squash merge, that isn't a problem.